[WebGPU] Minimum binding size for validation needs to be max of user specified and value in WGSL source

mwyrzykowski opened the pull request:
#### 24c8f5f321a0569ad1d02ebb3d45675af50e9539
<pre>
[WebGPU] Minimum binding size for validation needs to be max of user specified and value in WGSL source
<a href="https://bugs.webkit.org/show_bug.cgi?id=281057">https://bugs.webkit.org/show_bug.cgi?id=281057</a>
<a href="https://rdar.apple.com/137262987">rdar://137262987</a>

Reviewed by Tadeu Zagallo.

Ensure entrySize is at least the size of the minimum binding size
computed in the shader parsing.

* LayoutTests/fast/webgpu/nocrash/repro-281057-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/repro-281057.html: Added.
Add regression test.

* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::makeBindGroupEntryUsageData):
(WebGPU::Device::createBindGroup):
* Source/WebGPU/WebGPU/BindableResource.h:
* Source/WebGPU/WebGPU/Pipeline.mm:
(WebGPU::errorValidatingBindGroup):

Canonical link: <a href="https://commits.webkit.org/284987@main">https://commits.webkit.org/284987@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/55720aae4b5173e6b03cb70415f6f81fe6f14e15

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70954 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50366 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23727 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75058 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22159 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58165 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21977 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56123 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14605 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74020 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45764 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61166 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36571 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42419 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18600 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20500 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64359 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18961 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76774 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15186 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18162 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63863 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15230 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61225 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63820 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15742 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11913 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5557 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46167 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47239 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48522 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46981 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

tadeuzagallo commented on Source/WebGPU/WebGPU/Pipeline.mm:
```diff
@@ -191,16 +191,16 @@
+            if (mininumBufferSizes) {
                 if (auto bufferSizeIt = mininumBufferSizes->find(it->value.binding); bufferSizeIt != mininumBufferSizes->end())
-                    bufferSize = bufferSizeIt->value;
+                    bufferSize = std::max(bufferSize, bufferSizeIt->value);
```
I'm not sure if I'm interpreting the spec correctly, but shouldn't it be an error if `bufferSizeIt->value > bufferSize`? From the [spec](https://www.w3.org/TR/webgpu/#dom-gpubufferbindinglayout-minbindingsize):

> If this is not 0, pipeline creation additionally [validates](https://www.w3.org/TR/webgpu/#abstract-opdef-validating-shader-binding) that this value ≥ the [minimum buffer binding size](https://www.w3.org/TR/webgpu/#minimum-buffer-binding-size) of the variable.

mwyrzykowski replied:
oh nice observation, I should add that condition too. Will update the PR

mwyrzykowski committed:
[WebGPU] Minimum binding size for validation needs to be max of user specified and value in WGSL source
https://bugs.webkit.org/show_bug.cgi?id=281057
rdar://137262987

Reviewed by Tadeu Zagallo.

Ensure entrySize is at least the size of the minimum binding size
computed in the shader parsing.

* LayoutTests/fast/webgpu/nocrash/repro-281057-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/repro-281057.html: Added.
Add regression test.

* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::makeBindGroupEntryUsageData):
(WebGPU::Device::createBindGroup):
* Source/WebGPU/WebGPU/BindableResource.h:
* Source/WebGPU/WebGPU/Pipeline.mm:
(WebGPU::errorValidatingBindGroup):

Canonical link: https://commits.webkit.org/284987@main
